pkg/azureutils/logger.go

```diff
@@ -0,0 +1,189 @@
+	requestID := GetRequestID(obj)
+	startTimeStr := getAnnotationFromRuntimeObject(obj, consts.RequestStartimeKey)
+	var startTime time.Time
+	if startTimeStr != "" {
+		startTime, _ = time.Parse(consts.RequestTimeFormat, startTimeStr)
+	}
+	workflow := Workflow{
+		requestID: requestID,
+		startTime: startTime,
+		logger:    Logger{klogr.New().WithValues(append(GetObjectDetails(obj), consts.RequestIDKey, requestID)...)},
+	}
+	workflow.rootWorkflow = &workflow
+	ctx = workflow.SaveToContext(ctx)
+	return ctx, workflow
+}
+
+func GetObjectDetails(obj runtime.Object) []interface{} {
+	switch target := obj.(type) {
+	case *diskv1beta1.AzVolume:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.PvNameKey, target.Spec.PersistentVolume}
+		}
+	case *diskv1beta1.AzVolumeAttachment:
+		if target != nil {
+			return []interface{}{consts.VolumeNameLabel, target.Spec.VolumeName, consts.NodeNameLabel, target.Spec.NodeName, consts.RoleLabel, target.Spec.RequestedRole}
+		}
+	}
+	return []interface{}{}
+}
+
+func StartWorkflow(ctx context.Context, name string, details ...interface{}) (context.Context, Workflow) {
+	w := Workflow{
+		requestID: GenerateRequestID(),
+	}
+	w.logger = Logger{klogr.New().WithValues(consts.RequestIDKey, w.requestID).WithValues(details...)}
+	w.rootWorkflow = &w
+	return w.StartChildWorkflow(ctx, name)
+}
+
+func (w Workflow) StartChildWorkflow(ctx context.Context, name string) (context.Context, Workflow) {
+	childWorkflow := Workflow{
+		rootWorkflow: w.rootWorkflow,
+		startTime:    time.Now(),
+		requestID:    w.requestID,
+		logger:       w.logger,
+		name:         name,
+	}
+	return childWorkflow.SaveToContext(ctx), childWorkflow
+}
+
+func emptyWorkflow() Workflow {
+	w := Workflow{
+		startTime: time.Now(),
+		logger:    Logger{klogr.New()},
+	}
+	w.rootWorkflow = &w
+	return w
+}
+
+func (w Workflow) Finish(err error) {
+	latency := time.Since(w.startTime)
+	w.logger.WithValues(consts.Latency, latency, "error", err, "workflow", w.name).WithCallDepth(1).Info("Workflow completed.")
+}
+
+func (w Workflow) AddRequestToObject(obj runtime.Object) {
+	meta, _ := meta.Accessor(obj)
+	if meta == nil {
+		return
+	}
+	annotations := meta.GetAnnotations()
+	if annotations == nil {
+		annotations = map[string]string{}
+	}
+	annotations[consts.RequestIDKey] = w.requestID
+	annotations[consts.RequestStartimeKey] = w.startTime.Format(consts.RequestTimeFormat)
+	meta.SetAnnotations(annotations)
+}
+
+func (w Workflow) StartTime() time.Time {
+	return w.startTime
+}
+
+func (w Workflow) RequestID() string {
+	return w.requestID
+}
+
+func (w Workflow) Logger() Logger {
+	return w.logger
+}
+
+func (w Workflow) Name() string {
+	return w.name
+}
+
+func (w Workflow) AddDetailToLogger(details ...interface{}) {
+	w.logger = Logger{w.logger.WithValues(details...)}
+}
+
+func (w Workflow) SaveToContext(ctx context.Context) context.Context {
+	return context.WithValue(ctx, workflowKey, &w)
```

Comment:
Consider saving the logger to the `Context` as well so that logr/klogr clients will can log through our logger if they're context-aware, e.g.

```go
return logr.NewContext(context.WithValue(ctx, workflowKey, &w), w.logger)
```